net/udprelay/server.go

```diff
@@ -0,0 +1,517 @@
+			copy(m.Challenge[:], e.challenge[senderIndex][:])
+			reply := make([]byte, packet.GeneveFixedHeaderLength, 512)
+			gh := packet.GeneveHeader{Control: true, VNI: e.vni, Protocol: packet.GeneveProtocolDisco}
+			err := gh.Encode(reply)
+			if err != nil {
+				return
+			}
+			reply = append(reply, disco.Magic...)
+			reply = serverDisco.AppendTo(reply)
+			box := e.discoSharedSecrets[senderIndex].Seal(m.AppendMarshal(nil))
+			reply = append(reply, box...)
+			uw.WriteMsgUDPAddrPort(reply, nil, from)
+			// set new state
+			e.handshakeState[senderIndex] = disco.BindUDPRelayHandshakeStateChallengeSent
+			return
+		default:
+			// disco.BindUDPRelayEndpoint is unexpected in all other handshake states
+			return
+		}
+	case *disco.BindUDPRelayEndpointAnswer:
+		switch handshakeState {
+		case disco.BindUDPRelayHandshakeStateChallengeSent:
+			if from != e.addrPorts[senderIndex] {
+				// sender source has changed
+				return
+			}
+			if !bytes.Equal(discoMsg.Answer[:], e.challenge[senderIndex][:]) {
+				// bad answer
+				return
+			}
+			// sender is now bound
+			// TODO: Consider installing a fast path via netfilter or similar to
+			// relay (NAT) data packets for this serverEndpoint.
+			e.handshakeState[senderIndex] = disco.BindUDPRelayHandshakeStateAnswerReceived
+			// record last seen as bound time
+			e.lastSeen[senderIndex] = time.Now()
+			return
+		default:
+			// disco.BindUDPRelayEndpointAnswer is unexpected in all other handshake
+			// states, or we've already handled it
+			return
+		}
+	default:
+		// unexpected Disco message type
+		return
+	}
+}
+
+func (e *serverEndpoint) handleSealedDiscoControlMsg(from netip.AddrPort, b []byte, uw udpWriter, serverDisco key.DiscoPublic) {
+	senderRaw, isDiscoMsg := disco.Source(b)
+	if !isDiscoMsg {
+		// Not a Disco message
+		return
+	}
+	sender := key.DiscoPublicFromRaw32(mem.B(senderRaw))
+	senderIndex := -1
+	switch {
+	case sender.Compare(e.discoPubKeys[0]) == 0:
+		senderIndex = 0
+	case sender.Compare(e.discoPubKeys[1]) == 0:
+		senderIndex = 1
+	default:
+		// unknown Disco public key
+		return
+	}
+
+	const headerLen = len(disco.Magic) + key.DiscoPublicRawLen
+	discoPayload, ok := e.discoSharedSecrets[senderIndex].Open(b[headerLen:])
+	if !ok {
+		// unable to decrypt the Disco payload
+		return
+	}
+
+	discoMsg, err := disco.Parse(discoPayload)
+	if err != nil {
+		// unable to parse the Disco payload
+		return
+	}
+
+	e.handleDiscoControlMsg(from, senderIndex, discoMsg, uw, serverDisco)
+}
+
+type udpWriter interface {
+	WriteMsgUDPAddrPort(b []byte, oob []byte, addr netip.AddrPort) (n, oobn int, err error)
+}
+
+func (e *serverEndpoint) handlePacket(from netip.AddrPort, gh packet.GeneveHeader, b []byte, uw udpWriter, serverDisco key.DiscoPublic) {
+	if !gh.Control {
+		if !e.isBound() {
+			// not a control packet, but serverEndpoint isn't bound
+			return
+		}
+		var to netip.AddrPort
+		switch {
+		case from == e.addrPorts[0]:
+			e.lastSeen[0] = time.Now()
```

Comment:
nit: consider mono.Now for this use case

Reply:
Added a TODO in https://github.com/tailscale/tailscale/pull/15480/commits/4b930fa743abe3a4e39fdd96e772813dc4b8f4cd, and the thought had crossed my mind, but I'm being pedantic in terms of simpleness/correctness first before optimizing much.